src/customer/strategies/apple-pay/apple-pay-customer-strategy.ts

```diff
@@ -0,0 +1,411 @@
+        const applePaySession = this._sessionFactory.create(request);
+        this._handleApplePayEvents(applePaySession, this._paymentMethod, config);
+
+        applePaySession.begin();
+    }
+
+    private _getBaseRequest(
+        cart: Cart,
+        checkout: Checkout,
+        config: StoreConfig,
+        paymentMethod: PaymentMethod
+    ): ApplePayJS.ApplePayPaymentRequest {
+        const { storeProfile: { storeCountryCode, storeName } } = config;
+        const { currency: { code, decimalPlaces} } = cart;
+
+        const { initializationData : { merchantCapabilities, supportedNetworks } } = paymentMethod;
+
+        const requiresShipping = cart.lineItems.physicalItems.length > 0;
+        const total: ApplePayJS.ApplePayLineItem = requiresShipping ? {
+            label: storeName,
+            amount: `${checkout.grandTotal.toFixed(decimalPlaces)}`,
+            type: 'pending',
+        } : {
+            label: storeName,
+            amount: `${checkout.grandTotal.toFixed(decimalPlaces)}`,
+            type: 'final',
+        };
+
+        const request: ApplePayJS.ApplePayPaymentRequest = {
+            requiredBillingContactFields: ['postalAddress'],
+            requiredShippingContactFields: ['email', 'phone'],
+            countryCode: storeCountryCode,
+            currencyCode: code,
+            merchantCapabilities,
+            supportedNetworks,
+            lineItems: [],
+            total,
+        };
+
+        if (requiresShipping) {
+            request.requiredShippingContactFields?.push('postalAddress');
+        } else {
+            const lineItems: ApplePayJS.ApplePayLineItem[] = [
+                { label: this._subTotalLabel, amount: `${checkout.subtotal.toFixed(decimalPlaces)}`},
+            ];
+
+            checkout.taxes.forEach(tax =>
+                lineItems.push({ label: tax.name, amount: `${tax.amount.toFixed(decimalPlaces)}` }));
+
+            request.lineItems = lineItems;
+        }
+
+        return request;
+    }
+
+    private _handleApplePayEvents(
+        applePaySession: ApplePaySession,
+        paymentMethod: PaymentMethod,
+        config: StoreConfig
+    ) {
+        applePaySession.onvalidatemerchant = async event => {
+            try {
+                const { body: merchantSession } = await this._onValidateMerchant(paymentMethod, event);
+                applePaySession.completeMerchantValidation(merchantSession);
+            } catch (err) {
+                this._onError(err);
+            }
+        };
+
+        applePaySession.onshippingcontactselected = async event =>
+            this._handleShippingContactSelected(applePaySession, config, event);
+
+        applePaySession.onshippingmethodselected = async event =>
+            this._handleShippingMethodSelected(applePaySession, config, event);
+
+        applePaySession.oncancel = async () => {
+            await this._store.dispatch(this._remoteCheckoutActionCreator.signOut(paymentMethod.id));
```

Comment:
What happens if these async calls fail in this callback? Should you trigger the `onError` handler, like what you're doing with `onvalidatemerchant`?